viewing_party/party.py

```diff
@@ -1,23 +1,181 @@
+    user_watched_set = set(user_watched)
+    friends_watched_set = set(friends_watched)
+
+    unique_movies_watched_friends_only = friends_watched_set - user_watched_set
+
+    seen_titles = []
+    unique_movies_watched_friends_only_list = []
+
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            if movie["title"] in unique_movies_watched_friends_only and movie["title"] not in seen_titles:
+                unique_movies_watched_friends_only_list.append(movie)
+                seen_titles.append(movie["title"])
+
+    return unique_movies_watched_friends_only_list
 
-        
-# -----------------------------------------
 # ------------- WAVE 4 --------------------
-# -----------------------------------------
+def get_available_recs(user_data):
+    user_watched = []
+    for movie in user_data["watched"]:
+        user_watched.append(movie["title"])
+
+    subscriptions = user_data["subscriptions"]
+
+    friends_watched_movies = []
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            if movie not in friends_watched_movies:
+                friends_watched_movies.append(movie)
+
+    rec_movies = []
+    for movie in friends_watched_movies:
+        if movie["title"] not in user_watched and movie["host"] in subscriptions:
+            rec_movies.append(movie)
+
+    return rec_movies
 
-# -----------------------------------------
 # ------------- WAVE 5 --------------------
-# -----------------------------------------
+def get_new_rec_by_genre(user_data):
+    user_watched = []
+    for movie in user_data["watched"]:
+        user_watched.append(movie["title"])
+
+    friends_watched_movies = []
+
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            if movie not in friends_watched_movies:
+                friends_watched_movies.append(movie)
+
+    genre_count = {}
+    most_watched_genre = ""
+    if not user_data["watched"]:
+        return []
+
+    for movie in user_data["watched"]:
+        if movie["genre"] not in genre_count:
+            genre_count[movie["genre"]] = 1
+        else:
+            genre_count[movie["genre"]] += 1
+    most_watched_genre = max(genre_count, key=genre_count.get)
+
+    genre_recommendations = []
+    for movie in friends_watched_movies:
+        if movie["title"] not in user_watched and movie["genre"] == most_watched_genre:
+            genre_recommendations.append(movie)
+
+    return genre_recommendations
+
+def get_rec_from_favorites(user_data):
+    friends_watched = []
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            friends_watched.append(movie["title"])
+
+    user_watched = []
+    for movie in user_data["watched"]:
+        user_watched.append(movie["title"])
+
+    fav_recommendations = []
+    for movie in user_data["favorites"]:
+        if movie["title"] not in friends_watched:
+            fav_recommendations.append(movie)

```

Comment:
Same as before! Overall, y'all's logic is solid, just make sure to reuse what we've already written!